lib/ansible/module_utils/api.py

```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
     return wrapper
 
 
-def retry(retries=None, retry_pause=1):
-    """Retry decorator"""
-    def wrapper(f):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
+        return raise_error
+    return function_wrapper
 
-        def retried(*args, **kwargs):
-            retry_count = 0
-            if retries is not None:
-                ret = None
-                while True:
-                    retry_count += 1
-                    if retry_count >= retries:
-                        raise Exception("Retry limit exceeded: %d" % retries)
-                    try:
-                        ret = f(*args, **kwargs)
-                    except Exception:
-                        pass
-                    if ret:
-                        break
-                    time.sleep(retry_pause)
-                return ret
-
-        return retried
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
+    This assumes the function has already been called, so if retries == 1, the maximum attempts
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
```

Comment:
This is clever but it's broken my brain until I parsed it :)

The general maintainability advice (most linters actually enforce) is to pass lambdas as arguments when calling other functions (like passing predicates to filters and stuff).

Please move this to a named function. At least expand the outer function so it could be used to communicate it purpose.
```python
def noop(_func):
    return lambda *_args, **_kwargs: None
```

`return noop` reads better, doesn't it?

Reply:
I'm leaning towards just removing the changes to `retry`. It can be refactored to use the common function, but I just feel like it doesn't gain much of anything.

Reply:
Yeah, that'd work too. I'm just trying to give you a fresh-look perspective here...

Reply:
Removed the refactor.